### ISSUE TITLE
pexpect has very poor performance when searching large output streams 

### ISSUE THREAD

**Issue by ryanpetrello** [pexpect/pexpect#438]
Even with `searchwindowsize` set, pexpect's read/search algorithm seems to have _very poor_ performance for processes that write large quantities of data to stdout or stderr.  It gets significantly worse as the size of the output stream moves upwards.

`cat expect.py`
```python
#!/usr/bin/env python
import subprocess
import expect
# Write a ~5MB file
subprocess.call('base64 /dev/urandom | head -c 5000000 > /tmp/large', shell=True)
child = pexpect.spawn(
    'cat', ['/tmp/large'], [], env={}, ignore_sighup=True,
    encoding='utf-8', echo=False
)
while child.isalive():
    child.expect([], timeout=10000000, searchwindowsize=100)
```

When `cat /tmp/large` is run outside of pexpect on moderate hardware, it finishes (and prints) in seconds:

```
$ time cat /tmp/large
...
real	0m1.011s
user	0m0.000s
sys	0m0.130s
```

When run with pexpect, it takes notably longer:

```
time python expect.py
real	0m5.451s
user	0m1.970s
sys	0m2.720s
```

This timing gets slower and slower with larger data sets:
```
time python expect.10mb.py
real	0m24.363s
user	0m8.750s
sys	0m14.540s
```

```
time python expect.25mb.py
real	2m13.992s
user	0m46.930s
sys	1m21.810s
```

Anything beyond 50MB is almost intolerably slow - taking 30 minutes+ to finish on my cir. 2015 Macbook Pro (while finishing in < 30 seconds when run by hand in a terminal emulator).

`pexpect` has a configurable `maxread` that can be used to increase the number of bytes read at a time:
https://github.com/pexpect/pexpect/blob/master/pexpect/pty_spawn.py#L34

...but the pty seems to impose an OS-dependent upper limit (looks like 2k on my Macbook Pro), so simply bumping `maxread` doesn't help.

**Comment by ryanpetrello** [comment]
These two issues seem to be very closely related, and are possibly duplicates:

https://github.com/pexpect/pexpect/issues/385
https://github.com/pexpect/pexpect/issues/352

**Comment by ryanpetrello** [comment]
I passed this through cProfile and found that almost all of the time spent seems to be in the new data consumption/search:

![29218395-b5b7ebf8-7e82-11e7-9eca-45438551614c](https://user-images.githubusercontent.com/214912/29222480-a65151aa-7e90-11e7-9dd7-39584de015c2.png)

**Comment by ryanpetrello** [comment]
I originally thought that the problem was in the search algorithm, but that's not it, because my examples don't actually specify any search regular expressions. The problem is the way that string building and searching happens here:

https://github.com/pexpect/pexpect/blob/master/pexpect/expect.py#L17

Every iteration through this loop - which occurs every 2000 bytes by default - allocates a _new_ string.  So for a 50MB file, you end up allocating around 25,0000 new strings of increasing length:

`malloc 2k, malloc 4k, malloc 6k, ... malloc 50mb`

One solution to this is to *not* use immutable strings for a buffer here, but something like a mutable `StringIO`.  This diff is *not* complete (it needs to be adjusted to actually handle searching), but for examples like mine where there are no search strings, you can see that it's *much* more performant:

```diff
diff --git a/pexpect/expect.py b/pexpect/expect.py
index 660cfb5..3b1ceee 100644
--- a/pexpect/expect.py
+++ b/pexpect/expect.py
@@ -1,3 +1,4 @@
+from StringIO import StringIO
 import time
 
 from .exceptions import EOF, TIMEOUT
@@ -14,9 +15,9 @@ class Expecter(object):
         spawn = self.spawn
         searcher = self.searcher
 
-        incoming = spawn.buffer + data
-        freshlen = len(data)
-        index = searcher.search(incoming, freshlen, self.searchwindowsize)
+        spawn.buffer.write(data)
+        freshlen = spawn.buffer.len
+        index = searcher.search(spawn.buffer, freshlen, self.searchwindowsize)
         if index >= 0:
             spawn.buffer = incoming[searcher.end:]
             spawn.before = incoming[: searcher.start]
@@ -26,14 +27,12 @@ class Expecter(object):
             # Found a match
             return index
     
-        spawn.buffer = incoming
-    
     def eof(self, err=None):
         spawn = self.spawn
         from . import EOF
 
         spawn.before = spawn.buffer
-        spawn.buffer = spawn.string_type()
+        spawn.buffer = StringIO()
         spawn.after = EOF
         index = self.searcher.eof_index
         if index >= 0:
@@ -85,8 +84,8 @@ class Expecter(object):
             end_time = time.time() + timeout
 
         try:
-            incoming = spawn.buffer
-            spawn.buffer = spawn.string_type()  # Treat buffer as new data
+            incoming = spawn.buffer.getvalue()
+            spawn.buffer = StringIO()
             while True:
                 idx = self.new_data(incoming)
                 # Keep reading until exception or return.
@@ -279,10 +278,10 @@ class searcher_re(object):
         first_match = None
         # 'freshlen' doesn't help here -- we cannot predict the
         # length of a match, and the re module provides no help.
-        if searchwindowsize is None:
-            searchstart = 0
-        else:
-            searchstart = max(0, len(buffer) - searchwindowsize)
+        #if searchwindowsize is None:
+        #    searchstart = 0
+        #else:
+        #    searchstart = 0#max(0, len(buffer) - searchwindowsize)
         for index, s in self._searches:
             match = s.search(buffer, searchstart)
             if match is None:
diff --git a/pexpect/pty_spawn.py b/pexpect/pty_spawn.py
index 0d44230..669330a 100644
--- a/pexpect/pty_spawn.py
+++ b/pexpect/pty_spawn.py
@@ -1,3 +1,4 @@
+from StringIO import StringIO
 import os
 import sys
 import time
@@ -206,9 +207,9 @@ class spawn(SpawnBase):
         s.append('command: ' + str(self.command))
         s.append('args: %r' % (self.args,))
         s.append('buffer (last 100 chars): %r' % (
-                self.buffer[-100:] if self.buffer else self.buffer,))
+                self.buffer.getvalue()[-100:] if self.buffer else self.buffer,))
         s.append('before (last 100 chars): %r' % (
-                self.before[-100:] if self.before else self.before,))
+                self.before.getvalue()[-100:] if self.before else self.before,))
         s.append('after: %r' % (self.after,))
         s.append('match: %r' % (self.match,))
         s.append('match_index: ' + str(self.match_index))
@@ -737,7 +738,7 @@ class spawn(SpawnBase):
         # Flush the buffer.
         self.write_to_stdout(self.buffer)
         self.stdout.flush()
-        self.buffer = self.string_type()
+        self.buffer = StringIO()
         mode = tty.tcgetattr(self.STDIN_FILENO)
         tty.setraw(self.STDIN_FILENO)
         if escape_character is not None and PY3:
diff --git a/pexpect/spawnbase.py b/pexpect/spawnbase.py
index cb64582..ee0b828 100644
--- a/pexpect/spawnbase.py
+++ b/pexpect/spawnbase.py
@@ -1,3 +1,4 @@
+from StringIO import StringIO
 import codecs
 import os
 import sys
@@ -58,7 +59,7 @@ class SpawnBase(object):
         # max bytes to read at one time into buffer
         self.maxread = maxread
         # This is the read buffer. See maxread.
-        self.buffer = bytes() if (encoding is None) else text_type()
+        self.buffer = StringIO()
         # Data before searchwindowsize point is preserved, but not searched.
         self.searchwindowsize = searchwindowsize
         # Delay used before sending data to child. Time in seconds.
```

```
$ du -h /tmp/large
48M	/tmp/large
$ time python expect.py
...
real	0m7.832s
user	0m1.080s
sys	0m1.640s
```

**Comment by ryanpetrello** [comment]
Here's a more reproducible test case that illustrates the problem:

```diff
diff --git a/tests/test_performance.py b/tests/test_performance.py
index 7be0cf6..2cdb269 100755
--- a/tests/test_performance.py
+++ b/tests/test_performance.py
@@ -23,6 +23,7 @@ from __future__ import print_function
 import unittest, time, sys
 import platform
 import pexpect
+import re
 from . import PexpectTestCase
 
 # This isn't exactly a unit test, but it fits in nicely with the rest of the tests.
@@ -101,6 +102,11 @@ class PerformanceTestCase (PexpectTestCase.PexpectTestCase):
         self.faster_range(100000)
         print("100000 calls to faster_range:", (time.time() - start_time))
 
+    def test_large_stdout_stream(self):
+        e = pexpect.spawn('openssl rand -base64 {}'.format(1024*1024*3), timeout=2000)
+        resp = e.expect([re.compile(r'Password:\s*?$', re.M), pexpect.EOF, pexpect.TIMEOUT])
+        assert resp == 1  # index 1 == EOF
+
 if __name__ == "__main__":
     unittest.main()
```

```shell
time py.test tests/test_performance.py -k large_stdout_stream
12.89s user 2.70s system 94% cpu 16.433 total
```

**Comment by takluyver** [comment]
@ryanpetrello I take it this can be closed thanks to your PR #464. Let me know if there are more changes needed.
